elastic/kibana
# [Fleet] Update Package Policy UI to Support Upgrading Package Policies

## Summary

This PR encapsulates the "Upgrade package policy UI" milestone laid out in #106048. It adds support for upgrading package policies to the latest installed version of a given integration.

See [Figma Designs](https://www.figma.com/file/RVYNk5W3gyJdWOGDj8ixPH/Fleet-support?node-id=1016%3A125719)

<details>
<summary>Screen Recording</summary>

![Kapture 2021-08-04 at 08 20 42](https://user-images.githubusercontent.com/6766512/128180044-b6c1e0a9-5e11-466d-92cb-0c334cb114ba.gif)
</details>

- [x] Adjust `Fleet > Agent policies > Agent policy details` integrations table to match 
[design](https://user-images.githubusercontent.com/1965714/126019772-936f94a2-9242-426c-a9c3-2b000829e87b.png)
  - [x] Remove description column, move description to a help icon with tooltip (only if the package policy has a description)
  - [x] Add warning icon with tooltip if the policy package version =/= the currently installed package version
  - [x] Add `Upgrade policy` action menu item if the policy package version =/= the currently installed package version
- [x] Have the same changes in `Integrations > Integration details > Policies` table
  - [x] Add warning icon with tooltip if the policy package version =/= the currently installed package version
  - [x] Add `Upgrade policy` action menu item if the policy package version =/= the currently installed package version
- [x] When user clicks `Upgrade policy`, take them to the policy editor
  - [x] Perform a dry run against the upgrade package policies API 
  - [x] If the dry run comes back successful, add callout matching this [design](https://user-images.githubusercontent.com/1965714/126020107-1b480ab4-15dd-4327-a506-c8687c106216.png)
    - [x] Page title change for upgrade flow
    - [x] Page description change for upgrade flow
    - [x] Success callout
  - [x] If the dry run comes back with errors, add callout matching this [design](https://user-images.githubusercontent.com/1965714/126020320-b79ea895-8bd6-4828-b3db-4f966e509edd.png)
    - [x] Page title & description change (same as above)
    - [x] Warning callout 
    - [x] The policy editor should be populated with default values from the new package version (same as if the user were to add this integration for the first time)
    - [x] Clicking `previous configuration` should open a flyout with the JSON value of the existing package policy to enable users to copy over previous values (`diff` field from API)
    - [x] Investigate if time allows: how much of the previous configuration can we copy over "smartly"? 
- [x] When user clicks `Save integration`, save the adjusted policy with the new package version on the package metadata
  - [x] Any upgraded package policies should also trigger an update on their parent agent policies 
  - [x] Redirect user back to wherever they came from (Integrations policies table, or Fleet agent policies integrations table)

## Testing Instructions

1. Run the https://github.com/elastic/package-registry project locally, and ensure your local Kibana instance is pointing at your local package registry via `xpack.fleet.registryUrl: http://localhost:8080` in your `kibana.dev.yml`. Also ensure you have https://github.com/elastic/integrations cloned locally and https://github.com/elastic/elastic-package installed.
2. Install the `apache` integration to your default agent policy
3. Manually bump the version (and optionally add some new stream/input variables) to the `apache` package in its `manifest.yml`
4. Build the `apache` package via `elastic-package build`
5. Copy the built contents of your new `apache` package version to your package registry's `build` directory, e.g. `cp -R ../integrations/build/integrations/apache/<new_version> ./build/package-storage/packages/apache`
6. Restart your local package registry
7. Install the latest version of `apache` via the `Integrations -> Manage -> Apache -> Settings` screen
8. Navigate to either the `Integrations -> Manage -> Apache -> Policies` or `Fleet -> Agent Policies -> Default Policy -> Integrations` screens
9. Click the "Upgrade" button next to the `apache` package policy, or select `Upgrade Policy` from the actions menu in the table
10. Make any changes to custom variables you may have added on the "Upgrade Integration" screen, then save to confirm
11. Note the version of the package and any variable changes have been persisted to your package policy
